FEXHeaderUtils/FEXHeaderUtils/Filesystem.h

```diff
@@ -14,4 +22,270 @@ namespace FHU::Filesystem {
+      return Path;
+    }
+
+    return Path.substr(LastSeparator + 1);
+  }
+
+  inline fextl::string ParentPath(const fextl::string &Path) {
+    auto LastSeparator = Path.rfind('/');
+    if (LastSeparator == fextl::string::npos) {
+      // No separator. Likely relative `.`, `..`, `<Application Name>`, or empty string.
+      return Path;
+    }
+
+    return Path.substr(0, LastSeparator);
+  }
+
+  inline bool IsRelative(const fextl::string &Path) {
+    return Path.starts_with('/');
+  }
+
+  enum class CopyOptions {
+    NONE,
+    SKIP_EXISTING,
+    OVERWRITE_EXISTING,
+  };
+
+  inline bool CopyFile(const fextl::string &From, const fextl::string &To, CopyOptions Options = CopyOptions::NONE) {
+    bool DestTested = false;
+    bool DestExists = false;
+    if (Options == CopyOptions::SKIP_EXISTING && Exists(To.c_str())) {
+      DestTested = true;
+      DestExists = Exists(To.c_str());
+      if (DestExists) {
+        // If the destination file exists already and the skip existing flag is set then
+        // return true without error.
+        return true;
+      }
+    }
+
+    if (!DestTested) {
+      DestTested = true;
+      DestExists = Exists(To.c_str());
+    }
+
+    if (Options == CopyOptions::OVERWRITE_EXISTING && DestExists) {
+      // If we are overwriting and the file exists then we want to use `sendfile` to overwrite
+      int SourceFD = open(From.c_str(), O_RDONLY | O_CLOEXEC);
+      if (SourceFD == -1) {
+        return false;
+      }
+
+      int DestinationFD = open(To.c_str(), O_WRONLY | O_CREAT | O_TRUNC, 0200);
+      if (DestinationFD == -1) {
+        close(SourceFD);
+        return false;
+      }
+
+      struct stat buf;
+      if (fstat(SourceFD, &buf) != 0) {
+        close(DestinationFD);
+        close(SourceFD);
+        return false;
+      }
+
+      // Set the destination permissions to the original source permissions.
+      if (fchmod(DestinationFD, buf.st_mode) != 0) {
+        close(DestinationFD);
+        close(SourceFD);
+        return false;
+      }
+      bool Result = sendfile(DestinationFD, SourceFD, nullptr, buf.st_size) == buf.st_size;
+      close(DestinationFD);
+      close(SourceFD);
+      return Result;
+    }
+
+    if (!DestExists) {
+      // If the destination doesn't exist then just use rename.
+      return rename(From.c_str(), To.c_str()) == 0;
+    }
+
+    return false;
+  }
+
+  inline fextl::string LexicallyNormal(const fextl::string &Path) {
+    const auto PathSize = Path.size();
+
+    // Early exit on empty paths.
+    if (PathSize == 0) {
+      return {};
+    }
+
+    // Count the number of separators up front
+    const auto SeparatorCount = std::count(Path.begin(), Path.end(), '/');
+
+    // Needs to be a list so iterators aren't invalidated while erasing.
+    // Maximum number of list regions will be the counted separators plus 2.
+    // Multiplied by two since it allocates * 2 the size (32-bytes per element)
```

Comment:
What does this line mean?

Reply:
Rewrote that bit since it is confusing. Is just making sure to allocate enough stack space to handle the list element size.

Reply:
Now all the information needed to understand is there, but it's very disconnected. It's also unclear which of the comments refers to which part of the code.

Suggested rephrase:
```
Use std::list to store path elements to avoid iterator invalidation on insert/erase.
The list is allocated on stack to be more optimal. The size is determined by the
maximum number of list objects (separator count plus 2) multiplied by the list
element size (32-bytes per element: the string_view itself and the prev/next pointers).
```

Reply:
Updated comment to rephrased.